### PR TITLE
[Op]: Added definition of ROIAlignRotated.

### PR DESCRIPTION
**Pull request by pkowalc1**
### Details:
 - Added ROIAlignRotated definition to core
 - Refactored tests of ROIAlign shape inference info - to use the same tests for ROIAlignRotated 

### Tickets:
 - *CVS-115297*

### CODE REVIEW

**Review by praasz — CHANGES_REQUESTED** [review]
It also should be added to:
https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp#L400

and add test file similar to:
https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/tests/unit/shape_inference_test/roi_align_shape_inference_test.cpp

For op tests for attributes similar to:
https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/core/tests/visitors/op/fake_convert.cpp

File: src/core/tests/type_prop/roi_align.cpp [hunk 1]
```diff
@@ -48,7 +66,7 @@ TYPED_TEST_P(ROIAlignTest, simple_shape_inference) {
     const auto rois = make_shared<Parameter>(element::f16, rois_shape);
     const auto batch_indices = make_shared<Parameter>(element::i16, batch_shape);
 
-    const auto op = this->make_op(data, rois, batch_indices, 2, 2, 1, 1.0f, TypeParam::PoolingMode::AVG);
+    const auto op = this->make_roi_op(data, rois, batch_indices, 2, 2);
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/tests/type_prop/roi_align.cpp:69]: Those tests basically check only shape inference. My idea was to reuse exactly the same tests for 3 roi operators, which use exactly the same shape inference function(after the refactoring, which is part of this mr).

Those operators have different ctor parameters - however only few of them have real impact on shape inference. So I have added make_roi_op that  takes only parameters that crucial to calculating shape_inferece, rest is set as default. I don't see any reason to make it more complicated then it is - and IMO your suggestion makes it more complicated then it is.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    const auto& indicies_et = op->get_input_element_type(2);
+
+    NODE_VALIDATION_CHECK(op,
+                          indicies_et.is_integral_number(),
+                          "The data type for batch indices is expected to be an integer. Got: ",
+                          indicies_et);
+}
+}  // namespace validate
+
+template <class OpType, class TShape, int ROI_SECOND_DIM, class TRShape = result_shape_t<TShape>>
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp]: This is solved in different way, but ROI_SECOND_DIM  is gone.

File: src/core/src/op/roi_align_rotated.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//#include "openvino/reference/roi_align_rotated.hpp"
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/src/op/roi_align_rotated.cpp]: Fixed.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    const auto rois_ps_rank = rois_ps.rank();
+    const auto input_ps_rank = input_ps.rank();
+    const auto batch_indices_ps_rank = batch_indices_ps.rank();
+
+    auto output_shapes = std::vector<TRShape>(1);
+    auto& out_shape = output_shapes.front();
+    out_shape.reserve(4);
+
+    NODE_VALIDATION_CHECK(op, input_ps_rank.compatible(4), "Expected a 4D tensor for the input data. Got: ", input_ps);
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:68]: Sorry, the code is not mine - I am trying to make minimal changes that are needed to add new op, not fix all the issues. There are 3 similar ROI operators, which share the same code, so I wanted to make it clear.

This particular code was moved from file roi_align_shape_inference.hpp, which has passed earlier review :)

File: src/core/src/op/roi_align_rotated.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+#include "openvino/op/roi_align_rotated.hpp"
+
+#include "itt.hpp"
+#include "roi_align_rotated_shape_inference.hpp"
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/src/op/roi_align_rotated.cpp:9]: Fixed.

File: src/core/tests/type_prop/roi_align.cpp [hunk 6]
```diff
@@ -5,30 +5,48 @@
...
 #include "openvino/opsets/opset11.hpp"
+#include "openvino/opsets/opset14.hpp"
 
 using namespace std;
 using namespace ov;
 using namespace ov::opset11;
 using namespace testing;
 
-template <typename T>
-class ROIAlignTest : public TypePropOpTest<T> {};
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/tests/type_prop/roi_align.cpp:15]: Marking some ctor parameters as default changes the design a bit. I see that convention is that either you set set all parameters in ctor, or you use setter and default ctor, so I didn't want to change it.

**Comment by pkowalc1** [comment]
> It also should be added to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp#L400

This pr aims to add only op, no implementation at all. (So only symbolic inference should be possible). Implementation will be added in another PR.

> and add test file similar to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/tests/unit/shape_inference_test/roi_align_shape_inference_test.cpp

I have modified that test. It also runs those tests for new op. 

> For op tests for attributes similar to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/core/tests/visitors/op/fake_convert.cpp

Not sure what you meant by it, such tests are already in roi_align_shape_inference_test.cpp

File: src/core/shape_inference/include/roi_align_shape_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+#include <openvino/op/util/roi_align_base.hpp>
```
Comment (praasz) [inline src/core/shape_inference/include/roi_align_shape_utils.hpp:6]:
```suggestion
#include "openvino/op/util/roi_align_base.hpp"
```

File: src/core/include/openvino/op/roi_align.hpp [hunk 8]
```diff
@@ -4,31 +4,22 @@
...
+class OPENVINO_API ROIAlign : public util::ROIAlignBase {
 public:
-    OPENVINO_OP("ROIAlign", "opset3", op::Op);
+    OPENVINO_OP("ROIAlign", "opset3", util::ROIAlignBase);
     enum class PoolingMode { AVG, MAX };
 
     ROIAlign() = default;
     /// \brief Constructs a ROIAlign node matching the ONNX ROIAlign specification
     ///
-    /// \param input           Input feature map {N, C, H, W}
```
Comment (praasz) [inline src/core/include/openvino/op/roi_align.hpp:23]: These parameters are still in ctor description can be unchanged.

File: src/core/src/op/roi_align_rotated.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+                                 const Output<Node>& rois,
+                                 const Output<Node>& batch_indices,
+                                 const int pooled_h,
+                                 const int pooled_w,
+                                 const int sampling_ratio,
+                                 const float spatial_scale,
+                                 const bool clockwise_mode)
+    : ROIAlignBase{input, rois, batch_indices, pooled_h, pooled_w, sampling_ratio, spatial_scale},
+      m_clockwise_mode{clockwise_mode} {
+    constructor_validate_and_infer_types();
```
Comment (praasz) [inline src/core/src/op/roi_align_rotated.cpp:22]: The validation of op is run twice once when ROIAlignBase and 2nd here.
Is better to not execute validation in base class because:
- to avoid double run of validation.
- with new op version there can be some collisions in validation like for new version some case can correct but for base class not but there is still need to use base to share common part.

One solution could be not override `constructor_validate_and_infer_types` but make it base as protected helper and call it only in wanted operators.

File: src/core/tests/type_prop/roi_align.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/tests/type_prop/roi_align.cpp:15]: Yes and this is reason why test should be in separate file. The ROIAlignRotated is not ROIAligned and these test are only for ROI aligned.

Easy way just make a copy + use new ctor parameters and also validate.
Some test can be extracted to separate fixture and in ROI align rotate there can be only create instance for some common shape inference dedicated test.

All test create operator in same way and consistency is important, there is no reason to make this operator in different way.

File: src/core/include/openvino/op/roi_align.hpp [hunk 10]
```diff
@@ -98,34 +62,20 @@ class OPENVINO_API ROIAlign : public Op {
...
 public:
-    OPENVINO_OP("ROIAlign", "opset9");
+    OPENVINO_OP("ROIAlign", "opset9", util::ROIAlignBase);
     enum class PoolingMode { AVG, MAX };
     enum class AlignedMode { ASYMMETRIC, HALF_PIXEL_FOR_NN, HALF_PIXEL };
 
     ROIAlign() = default;
     /// \brief Constructs a ROIAlign operation.
     ///
-    /// \param input           Input feature map {N, C, H, W}
```
Comment (praasz) [inline src/core/include/openvino/op/roi_align.hpp:119]: Restore comment

File: src/core/include/openvino/op/roi_align_rotated.hpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+/// \ingroup ov_ops_cpp_api
+class OPENVINO_API ROIAlignRotated : public util::ROIAlignBase {
+public:
+    OPENVINO_OP("ROIAlignRotated", "opset14", util::ROIAlignBase);
+
+    ROIAlignRotated() = default;
+    /// \brief Constructs a ROIAlignRotated operation.
+    ///
+    /// \param clockwise_mode  If true, rotation angle is interpreted as clockwise, otherwise as counterclockwise
+    ROIAlignRotated(const Output<Node>& input,
```
Comment (praasz) [inline src/core/include/openvino/op/roi_align_rotated.hpp:23]: Describe all parameters.

**Comment by praasz** [comment]
> It also should be added to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp#L400
> 
> and add test file similar to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/plugins/intel_cpu/tests/unit/shape_inference_test/roi_align_shape_inference_test.cpp
> 
> For op tests for attributes similar to: https://github.com/praasz/openvino/blob/7a1cd37b9db5b31b36ef1ae646bdedebdfbac261/src/core/tests/visitors/op/fake_convert.cpp

This is also part of basic op creation.

Operator should be added to shape inference factory and then test file should be created to test static shapes.
src/plugins/intel_cpu/tests/unit/shape_inference_test/roi_align_rotated_shape_inference_test.cpp

This will also build `shape_infer` for static shapes for new operator and detects any compilation issue (I think there will any).


In `src/core/tests/visitors/op/` folder add file for new op and test for visitors.

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/core/include/openvino/op/util/roi_align_base.hpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                 const Output<Node>& batch_indices,
+                 const int pooled_h,
+                 const int pooled_w,
+                 const int sampling_ratio,
+                 const float spatial_scale);
+
+public:
+    // Return size of second dimension of the ROI input.
+    // Needed for validation check.
+    virtual int get_rois_input_second_dim_size() const = 0;
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/util/roi_align_base.hpp:44]: I disagree, it is a part of the definition of the op.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    NODE_VALIDATION_CHECK(op, rois_ps_rank.compatible(2), "Expected a 2D tensor for the ROIs input. Got: ", rois_ps);
+    NODE_VALIDATION_CHECK(op,
+                          batch_indices_ps_rank.compatible(1),
+                          "Expected a 1D tensor for the batch indices input. Got: ",
+                          batch_indices_ps);
+
+    if (rois_ps_rank.is_static()) {
+        const auto& rois_second_dim = rois_ps[1];
+        NODE_VALIDATION_CHECK(op,
+                              rois_second_dim.compatible(op->get_rois_input_second_dim_size()),
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: It is a part of the definition - and it is needed in shape inference. Also it is a const getter.

File: src/core/include/openvino/op/util/roi_align_base.hpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    int get_pooled_h() const {
+        return m_pooled_h;
+    }
+
+    void set_pooled_h(const int h) {
+        m_pooled_h = h;
+    }
+
+    int get_pooled_w() const {
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/util/roi_align_base.hpp:57]: Sorry, but this is the convention. I have checked other op impl and everyone is doing it this way. I have also copied it from previous roi align impl. So I will stick to convention.

However, I don't like it as well - they should be however defined as inline in header file but outside the class def. But - as I said - I prefer to stick to the convention. If you want to change it, you can always make a new pr and fix all other ops impl - just to have single convention.

File: src/core/shape_inference/include/roi_align_shape_inference.hpp [hunk 15]
```diff
@@ -4,115 +4,23 @@
...
-
-    return output_shapes;
-}
-}  // namespace roi_align
 
 namespace v3 {
 template <class TShape, class TRShape = result_shape_t<TShape>>
 std::vector<TRShape> shape_infer(const ROIAlign* op, const std::vector<TShape>& input_shapes) {
-    return roi_align::shape_infer(op, input_shapes);
+    return roi_align::shape_infer<TShape, TRShape>(op, input_shapes);
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_inference.hpp:16]: I prefer to be explicit with templates in such cases.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: Shape inference is not enough to put it in public API. It can be part only of shape inference.

File: src/core/src/op/roi_align_rotated.cpp (hunk 9 above)
Comment (pkowalc1) [reply src/core/src/op/roi_align_rotated.cpp:22]: That was a mistake. fixed.

File: src/core/include/openvino/op/roi_align_rotated.hpp (hunk 11 above)
Comment (pkowalc1) [reply src/core/include/openvino/op/roi_align_rotated.hpp:23]: Those parameter are already described in a base. Will add a comment about that fact, since copying it is a wrong idea.

File: src/core/tests/type_prop/roi_align.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/tests/type_prop/roi_align.cpp:15]: I disagree - ROIAlign and ROIAlignRotated are very similar ops - to the extend that they could be a single op. There was a discussion about it and the conclusion was that it is better to have separate op definition, but maybe the implementation will be the same. 

In particular, those two ops have exactly the same shape_inference function. The difference between ROIAlignRotated and ROIAlign is similar to v3::ROIAlign andv9::ROIAlign. 

So the ROIAlignRotated should pass the same tests as the ROIAlign, becasue those tests only common functionalities - and I made it explicit. Maybe I can rename them to sth like ROIAlignCommonTest and then(if there will be ever a need), create additional file(s) that test specific functionalities of given op.  Copying (as you suggested) is the worst thing you can do here.

About consistency - please check tests like RNNSeqBaseTest - they follow exactly the same pattern as I did here: you have a common test for RNNSequence, GRUSequence and LSTMSequence.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: 1) This is actually need for validation, which is (incorrectly) done in shape_inference - so far you had a magic numbers there.
2) It is a part of op definition - please refer to *rst files. 
3) get_pooled_h() is also used only in tests and shape inference - is it enough to make it to public API?

File: src/core/include/openvino/op/roi_align.hpp (hunk 10 above)
Comment (pkowalc1) [reply src/core/include/openvino/op/roi_align.hpp:119]: Fixed in different way.

File: src/core/include/openvino/op/roi_align_rotated.hpp (hunk 11 above)
Comment (t-jankowski) [reply src/core/include/openvino/op/roi_align_rotated.hpp:24]: These comments are used for docs generation and in IDE helper pop-ups. The other, more important, reason to have it fully described is implicit inconsistency once the base class changes.
Please apply description of all parameters to both versions.

**Review by t-jankowski — DISMISSED** [review]

File: src/core/include/openvino/op/util/roi_align_base.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply src/core/include/openvino/op/util/roi_align_base.hpp:57]: This is not a convention. We can not refactor everything. Please stick to the @praasz advice.

[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/util/roi_align_base.hpp:57]: First of all, you guys commenting on the code which was not implemented by me. That code was just moved from file roi_align.hpp, as necessary step to introduce a common base for roi operators. It means that this code has passed the interview already. What is more, those functions where implemented 11 months ago by @praasz himself(it doesn't really matter, but it is worth to point it out here).

Now, if you guys want to introduce a new code convention every 11 month - please do it in separate PR and fix all the places. If you "can't refactor everything" - it is better to NOT refactor at all - otherwise you create a mess: 
- in this PR, because it is not about changing the code convention, is is about adding new operator def. Ideally, every PR should do only one logical thing. Changing code convention is completely orthogonal to adding new op def.
- in whole codebase, since nobody knows what is really the convention here. Every file might be different, just because somebody "had a time to fix this file or didn't have". 

Now, if you think that changing that is necessary, please follow up with a new pr that will fix the code to the single convention.

File: src/core/include/openvino/op/roi_align_rotated.hpp (hunk 11 above)
Comment (pkowalc1) [reply src/core/include/openvino/op/roi_align_rotated.hpp:24]: > The other, more important, reason to have it fully described is implicit inconsistency once the base class changes

Exactly - what you suggest here might introduce inconsistency, because you will have the same param described in a few places. Somebody might change the def of some param and forget to update the comments in all the paces. This simply breaks a single definition rule. Because of that, every param should be described only once and the base class it is where it should be done, since it is where those params are defined.

Having said that, if the problem is that the tool used for generation docs cannot handle it i might to do as you suggest as an exception, but that tool should be fixed as soon as possible. Otherwise I disagree with this comment.

File: src/core/include/openvino/op/util/roi_align_base.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/util/roi_align_base.hpp:44]: @jane-intel Please refer to my comment on very similar issue below - don't want to double the conversation - I have tried to motivate why I think that it is a part of the def. If you disagree, please motivate it.

**Review by mitruska — COMMENTED** [review]
In general the op class and shape inference code looks functionally correct. 
Good that common base class has been introduced and shape_infer code reused.

Let's resolve the minor comments, and add leftovers mentioned by @praasz.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: Apply solution propose by @mitruska or similar.

**Review by praasz — DISMISSED** [review]
The `get_rois_input_second_dim_size` is still in public API, apply e.g. solution proposed by @mitruska.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: Guys, what you propose is simply scattering the definition of op into multiple files. This is generally speaking bad idea. Solution proposed by @mitruska is exactly the same as it already was in roi_align - add magic number but put it into separate function in separate file called 'shape_inference', **while that magic number is not needed for shape inference at all - is needed only for validation, which is incorrectly mixed up with shape inference...**

The root problem is that pattern that shape inference was separated physically from op definition - which is IMO wrong design choice. Shape inference is integral part of any op and it should be kept together. Mixing it with validation is a even worse practice IMO, but I am not going to solve all of those issue here, as it is not related to the roi_align_rotated itself. 

However I was considering another solution - making get_rois_input_second_dim_size() private and moving validation to validate_and_infer_types() in ROIAlignBase, BUT it is a refactoring of current functionality and as such - should be done in separate pr, because it is not directly related to roi align rotated.

To sum up, what I proposed here is a kind of compromise between keeping everything in one place(as it should be) and NOT refactoring everything from ground up - because it is not directly related to the roi_align_rotated op itself. As such, I believe my proposition is better to yours and will keep it - honestly I don't see any drawbacks of having a const getter that defines a number directly from op specification and none of you were able so far to give me convincing explanation.

File: src/core/include/openvino/op/util/roi_align_base.hpp [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+private:
+    int m_pooled_h;
+    int m_pooled_w;
+    int m_sampling_ratio;
+    float m_spatial_scale;
+};
+
+}  // namespace util
+}  // namespace op
+}  // namespace ov
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/util/roi_align_base.hpp]: Fixed.

File: src/core/shape_inference/include/roi_align_shape_utils.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: There no need to add public member for such values or refactor the code.
The separation of shape inference has been done some time ago intentionally and consequence are known and accepted.

If there is possible to improve and make it more integrated, prepare ticket/PR with concept and if that give improvement
then it can be propagated to other operators.

[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/roi_align_shape_utils.hpp:78]: The need to keep such public members is a consequence of of the mixing validation and shape(along with separation shape inference) in one place. If you say that was done intentionally and consequences are known and accepted - then please do not block this PR, since once again - t**his is the one of the consequences of such design.**
